internal/service/worker/controller.go

```diff
@@ -31,39 +31,173 @@ func GrpcRegister() {
 }
 
 func (c *Controller) GetWorker(ctx context.Context, request *workerV1beta1.GetWorkerRequest) (*workerV1beta1.GetWorkerResponse, error) {
-	v, ok := c.states[request.WorkerId]
-	if !ok {
-		return nil, status.Errorf(codes.NotFound, "WorkerId: %s", request.WorkerId)
+	workerID := request.GetWorkerId()
+	var v validate.Validator
+	v.ValidateString("worker_id", workerID, validate.IsNotNull)
+
+	errorInfo := v.ErrorInfo()
+	if errorInfo != nil {
+		st, err := status.New(codes.InvalidArgument, "bad request").
+			WithDetails(errorInfo)
+		if err != nil {
+			panic(fmt.Errorf("enrich grpc status with details error: %w", err))
+		}
+		return nil, st.Err()
+	}
+
+	worker, has, err := GetWorker(ctx, request.WorkerId)
+
+	if err != nil {
+		st := status.Newf(codes.Internal, "get worker error: %s", err)
+		st, err = st.WithDetails(request)
+		if err != nil {
+			panic(fmt.Errorf("enrich grpc status with details error: %w", err))
+		}
+		return nil, st.Err()
+	}
+
+	if !has {
+		st := status.Newf(codes.NotFound, "worker not found")
+		st, err := st.WithDetails(request)
+		if err != nil {
+			panic(fmt.Errorf("enrich grpc status with details error: %w", err))
+		}
+		return nil, st.Err()
 	}
+
+	w := transformWorker(worker)
 	return &workerV1beta1.GetWorkerResponse{
-		Worker: &workerV1beta1.Worker{
-			WorkerId: request.WorkerId,
-			State:    v,
-		},
+		Worker: w,
 	}, nil
 }
-func (c *Controller) UpdateWorker(ctx context.Context, request *workerV1beta1.UpdateWorkerRequest) (*workerV1beta1.UpdateWorkerResponse, error) {
-	c.states[request.WorkerId] = request.GetState()
-	return &workerV1beta1.UpdateWorkerResponse{
-		Worker: &workerV1beta1.Worker{
-			WorkerId: request.GetWorkerId(),
-			State:    request.State,
-		},
+
+func (c *Controller) UpdateWorkerStatus(ctx context.Context, request *workerV1beta1.UpdateWorkerStatusRequest) (*workerV1beta1.UpdateWorkerStatusResponse, error) {
+	workerID := request.GetWorkerId()
+	workerStatus := request.GetStatus()
+	var v validate.Validator
+	v.ValidateString("worker_id", workerID, validate.IsNotNull)
+	v.Validate("worker_status", workerStatus, ValidateStatus)
+
+	errorInfo := v.ErrorInfo()
+	if errorInfo != nil {
+		st, err := status.New(codes.InvalidArgument, "bad request").
+			WithDetails(errorInfo)
+		if err != nil {
+			panic(fmt.Errorf("enrich grpc status with details error: %w", err))
+		}
+		return nil, st.Err()
+	}
+	s := transformStatusFrom(workerStatus)
+	err := SetWorkerStatus(ctx, workerID, s)
+
+	if err != nil {
+		st := status.Newf(codes.Internal, "update worker status error: %s", err)
+		st, err = st.WithDetails(request)
```

Comment:
I think I originally suggested we include the request in error details, which was a mistake on my part. The client already has the request / knows what they send.

Reply:
agree